### PR TITLE
Add support for storing isolated bloqs, like `GlobalPhaseGate`, in a `CompositeBloq`

### PR DESCRIPTION
**Pull request by tanujkhattar**
Fixes https://github.com/quantumlib/Qualtran/issues/422

Does the following:
1. Updates `CompositeBloq` to store both nodes and edges of the compute graph instead of only storing the nodes
2. Updates `cbloq_to_quimb` to add logic for adding isolated bloqs to the tensor networks
3. Updates `CirqGateAsBloq` to have an empty signature when `cirq.num_qubits(self.gate) == 0`. 
4. Updates `cirq_optree_to_cbloq` to correctly operations acting on 0 qubits. 
5. Tests the entire end to end workflow by wrapping `cirq.GlobalPhaseGate` in a bloq and using tensor contraction to verify that the resulting unitary is identical to the expected cirq unitary. 

This is needed for porting over hubbard model implementation from Cirq-FT to Qualtran.

### CODE REVIEW

**Review by mpharrigan — CHANGES_REQUESTED** [review]
this is a pretty fundamental change to our primary datastructure. I don't think we should support scalars as bloqs

**Comment by tanujkhattar** [comment]
What are your concerns? Why do you think we should not support scalars as Bloqs? 


Also, we can not just drop the global phase, so we need to come up with an alternative solution if you think this is not the right way. Do you have alternative proposals in mind?

**Comment by mpharrigan** [comment]
Some historical discussion:

https://github.com/quantumlib/Cirq/pull/4697

https://github.com/quantumlib/Cirq/pull/1743

----

I don't know what you mean by "drop the global phase". Every bloq and composite bloq currently has a well-defined global phase.

**Comment by tanujkhattar** [comment]
> I don't know what you mean by "drop the global phase". Every bloq and composite bloq currently has a well-defined global phase.

Okay, let me try to explain this differently. The use case is that often times it is possible that I want to implement a unitary `U` but, using the existing primitives, I have a way of implementing a unitary `Q` where `Q` and `U` are equal up to global phase. Having access to a global phase gate let's me define a decomposition of `U` in terms of simpler primitives + `GlobalPhaseGate`; which is convenient. It is important to ensure that the decomposition has the same global phase as `U`, or else the auto generated decomposition for a controlled invocation of `U` will result in a wrong unitary. 


Not having the ability to add `GlobalPhaseGate` would make it annoying for us to find ways to shoehorn this global phase into one of the other top level bloqs used in the decomposition; or use an additional qubit in `|1>` state and apply a `Z` gate to it -- both of these are wasteful / inconvenient.

File: qualtran/cirq_interop/_cirq_to_bloq.py [hunk 1]
```diff
@@ -80,7 +80,8 @@ def signature(self) -> 'Signature':
                     for reg in self.gate.signature
                 ]
             )
-        return Signature([Register('qubits', shape=cirq.num_qubits(self.gate), bitsize=1)])
+        shape = cirq.num_qubits(self.gate)
+        return Signature([Register('qubits', shape=shape, bitsize=1)]) if shape else Signature([])
```
Comment (mpharrigan) [inline qualtran/cirq_interop/_cirq_to_bloq.py:84]: should `shape` be called `n_qubits`? is it always an integer? If it is: the condition should be `if shape > 0`

File: qualtran/simulation/quimb_sim.py [hunk 2]
```diff
@@ -125,11 +127,13 @@ def _assign_outgoing(cxn: Connection) -> Soquet:
         if pos is not None:
             fix[tuple([binst])] = pos[binst]
 
-    inds = get_right_and_left_inds(cbloq.signature)
-    if len(inds) != 2:
-        # In this case, the tensor network is a vector (bra/ket) instead of a unitary operation.
-        return tn, fix
+    # Special case-1: Add all isolated Bloqs with no connections (eg: Global phase gates)
+    for binst in cbloq.bloq_instances:
+        print(f'DEBUG QUIMB: {binst=}')
```
Comment (mpharrigan) [inline qualtran/simulation/quimb_sim.py:132]: debug print

File: qualtran/simulation/quimb_sim.py [hunk 3]
```diff
@@ -104,8 +104,10 @@ def _assign_outgoing(cxn: Connection) -> Soquet:
             return cxn.right
         return cxn.left
 
+    visited_bloqs: Set[BloqInstance] = set()
     for binst, incoming, outgoing in cbloq.iter_bloqnections():
         bloq = binst.bloq
+        visited_bloqs.add(binst)
```
Comment (mpharrigan) [inline qualtran/simulation/quimb_sim.py:110]: I think the correct fix would be to modify `iter_bloqnections` to yield all the bloqs with empty `incoming` and `outgoing`.

**Comment by mpharrigan** [comment]
> I want to implement a unitary U but, using the existing primitives, I have a way of implementing a unitary Q where Q and U are equal up to global phase.

Ok, I'm open to this. We have other bookkeeping bloqs. 

I suspect this is still going to cause breakages elsewhere, but I guess we'll cross that bridge when we get there.

Does the drawing code work?

File: qualtran/simulation/quimb_sim.py (hunk 3 above)
Comment (tanujkhattar) [reply qualtran/simulation/quimb_sim.py:110]: done

File: qualtran/simulation/quimb_sim.py (hunk 2 above)
Comment (tanujkhattar) [reply qualtran/simulation/quimb_sim.py]: removed

File: qualtran/cirq_interop/_cirq_to_bloq.py (hunk 1 above)
Comment (tanujkhattar) [reply qualtran/cirq_interop/_cirq_to_bloq.py]: Yes, it's always an integer. Changed.

**Comment by tanujkhattar** [comment]
>I suspect this is still going to cause breakages elsewhere, but I guess we'll cross that bridge when we get there.

Yeah, the good thing is that all the tests pass and drawing code works (see below) and the quimb simulation works (as shown in the test). For whatever is remaining, we'll fix it as we discover it.

> Does the drawing code work? -

Yep! 

image

**Review by mpharrigan — APPROVED** [review]
Thanks for your patience. LGTM

**Pull request merged**